### PR TITLE
Changed AutoAck to AcknowledgmentMode

### PR DESCRIPTION
**Pull request by gingters**

### CODE REVIEW

File: Thinktecture.Relay.Server/OnPremise/IOnPremiseConnectorRequest.cs [hunk 1]
```diff
@@ -58,7 +58,7 @@ public interface IOnPremiseConnectorRequest
 		/// <summary>
 		/// Gets whether this request should be auto-acknowledged when taken from the RabbitMQ
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.Server/OnPremise/IOnPremiseConnectorRequest.cs]: This needs a new summary tag.

File: Thinktecture.Relay/Server/AcknowledgmentMode.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+namespace Thinktecture.Relay.Server
+{
+	public enum AcknowledgmentMode
+	{
+		/// <summary>
+		/// Requests in the message queue will be automatically acknowledged by
```
Comment (thomashilzendegen) [inline Thinktecture.Relay/Server/AcknowledgmentMode.cs]: `Requests in the message queue will be automatically acknowledged by the on-premise before contacting the local target`

File: Thinktecture.Relay/Server/AcknowledgmentMode.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+		/// the OnPremiseConnector before executing the request against the on premise api
+		/// </summary>
+		Default,
+
+		/// <summary>
+		/// Requests in the message queue will be automatically acknowledged by
```
Comment (thomashilzendegen) [inline Thinktecture.Relay/Server/AcknowledgmentMode.cs]: `Requests in the message queue will be automatically acknowledged by the relay server before sending to the on-premise`

File: Thinktecture.Relay/Server/AcknowledgmentMode.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+		/// the relay server before sending them to the OnPremiseConnector
+		/// </summary>
+		Auto,
+
+		/// <summary>
+		/// Requests in the message queue will not be acknowleged by the Relay system
```
Comment (thomashilzendegen) [inline Thinktecture.Relay/Server/AcknowledgmentMode.cs]: `Requests in the message queue need to be acknowledged manually`

File: Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs [hunk 5]
```diff
@@ -309,11 +306,20 @@ private void OnReconnecting()
 			}
 		}
 
-		private Task AcknowlegdeRequest(IOnPremiseTargetRequest request)
+		private Task AcknowlegdeRequestIfRequired(IOnPremiseTargetRequest request)
 		{
-			_logger?.Debug("Sending acknowlegde to relay server. request-id={RequestId}, acknowledge-id={AcknowledgeId}", request.RequestId, request.AcknowledgeId);
+			_logger.Verbose("Checking if acknowledged is required. request-id={RequestId}, acknowledgment-mode={AcknowledgmentMode}, acknowledge-id={AcknowledgeId}", request.RequestId, request.AcknowledgmentMode, request.AcknowledgeId);
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs]: Maybe this isn't really necessary to log?

File: Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs [hunk 6]
```diff
@@ -309,11 +306,20 @@ private void OnReconnecting()
...
+
+			if (request.AcknowledgmentMode == AcknowledgmentMode.Default && !String.IsNullOrEmpty(request.AcknowledgeId))
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs]: Maybe change to a `switch` which does nothing in the other modes, but the other modes are visible? In that case the first check could be for an existing `AcknowledgeId` and early exit when not.

File: Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs [hunk 7]
```diff
@@ -309,11 +306,20 @@ private void OnReconnecting()
...
+			{
+				_logger?.Debug("Sending acknowlegde to relay server. request-id={RequestId}, acknowledge-id={AcknowledgeId}", request.RequestId, request.AcknowledgeId);
+
+				return SendToRelayAsync($"/request/acknowledge?id={ConnectionId}&tag={request.AcknowledgeId}", HttpMethod.Get, null, null, CancellationToken.None);
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs]: Doing an `await` here means no `return` later (with `async` method of course).

File: Documentation/en/5-relayserver-interceptor-development.md [hunk 8]
```diff
@@ -45,7 +45,7 @@ The following values ​​can be changed:
 - `HttpHeaders`: HTTP headers can be removed, new added or existing changed.
 - `AlwaysSendToOnPremiseConnector`: Setting this to true will cause the request to be relayed to the OnPremiseConnector even if the interceptor immediately answers it by returning an `HttpResponseMessage`.
 - `Expiration`: The TTL of this request in the RabbitMQ can be changed here.
-- `AutoAcknowledge`: If set to true, the request will be automatically deleted from the RabbitMQ when read, if false, the RelayServer will wait for an OnPremiseConnector ACK before deleting the request from the message queue.
+- `AcknowledgmentMode`: This determines whether the OnPremiseConnector will acknowledge a request (Default), the RelayServer will auto-acknowledge the request (Auto), of if custom node on the target needs to manually Acknowledge the request (Manual). For manual acknowledgement, send an HTTP GET request to the `/request/acknowledge/` endpoint on the relay server and pass the query string parameters `tag` with the AcknowledgeId and `id` with the ConnectionId. This request needs to provide a valid bearer token in the Authorization header.
```
Comment (thomashilzendegen) [inline Documentation/en/5-relayserver-interceptor-development.md]: No slash at the end, isn't it?

File: Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/AcknowledgmentMode.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+namespace Thinktecture.Relay.OnPremiseConnector.OnPremiseTarget
+{
+	/// <summary>
+	/// This is the pendant for the server enum AcknowledgementMode and should be kept separate
+	/// </summary>
+	public enum AcknowledgmentMode
+	{
+		/// <summary>
+		/// Requests in the message queue will be automatically acknowledged by
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/AcknowledgmentMode.cs]: `Requests in the message queue will be automatically acknowledged by the on-premise before contacting the local target`

File: Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/AcknowledgmentMode.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+		/// the OnPremiseConnector before executing the request against the on premise api
+		/// </summary>
+		Default,
+
+		/// <summary>
+		/// Requests in the message queue will be automatically acknowledged by
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/AcknowledgmentMode.cs]: `Requests in the message queue will be automatically acknowledged by the relay server before sending to the on-premise`

File: Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/AcknowledgmentMode.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+		/// the relay server before sending them to the OnPremiseConnector
+		/// </summary>
+		Auto,
+
+		/// <summary>
+		/// Requests in the message queue will not be acknowleged by the Relay system
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/AcknowledgmentMode.cs]: `Requests in the message queue need to be acknowledged manually`

File: Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/IOnPremiseTargetRequest.cs [hunk 12]
```diff
@@ -56,5 +56,10 @@ public interface IOnPremiseTargetRequest
 		/// Gets the request stream if the body is too large (was requested by a second request)
 		/// </summary>
 		Stream Stream { get; }
+
+		/// <summary>
+		/// Gets the mode of Acknowledgment of this request
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/OnPremiseTarget/IOnPremiseTargetRequest.cs]: Should be lower-case `a`.

File: Thinktecture.Relay/Server/Interceptor/IInterceptedRequest.cs [hunk 13]
```diff
@@ -41,6 +41,6 @@ public interface IInterceptedRequest : IReadOnlyInterceptedRequest
 		/// Gets whether this request should be auto-acknowledged when taken from the RabbitMQ
```
Comment (thomashilzendegen) [inline Thinktecture.Relay/Server/Interceptor/IInterceptedRequest.cs]: This needs a new summary tag.

File: Documentation/en/5-relayserver-interceptor-development.md (hunk 8 above)
Comment (gingters) [reply Documentation/en/5-relayserver-interceptor-development.md]: In der Tat.

File: Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs (hunk 5 above)
Comment (gingters) [reply Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs]: That was very important during testing, so verbose is fine imho?

File: Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs [hunk 14]
```diff
@@ -306,20 +306,25 @@ private void OnReconnecting()
 			}
 		}
 
-		private Task AcknowlegdeRequestIfRequired(IOnPremiseTargetRequest request)
+		private async void AcknowlegdeRequestIfRequired(IOnPremiseTargetRequest request)
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs]: An `async void` is not recommended.

File: Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs [hunk 15]
```diff
@@ -258,7 +258,7 @@ private void OnReconnecting()
 				}
 				finally
 				{
-					await AcknowlegdeRequestIfRequired(request).ConfigureAwait(false);
+					AcknowlegdeRequestIfRequired(request);
```
Comment (thomashilzendegen) [inline Thinktecture.Relay.OnPremiseConnector/SignalR/RelayServerConnection.cs]: Keep `await` with `ConfigureAwait(false)`.

**Pull request merged**